### ISSUE TITLE
iOS mic recording has a clicking noise with the option .defaultToSpeaker = true

### ISSUE THREAD

**Issue by sihwapark** [AudioKit/AudioKit#1311]
On my iPhone 7, `.defaultToSpeaker = true` always causes clicking noise in a recorded sound but `.defaultToSpeack = false` doesn't. (When headphones are plugged and `.defaultToSpeaker = true`, there is no noise.). This issue also appears in the iOS Recorder example of AudioKit. 

My temporary solution is toggling `defaultToSpeaker` whenever I record or play back as below:
```
AKSettings.defaultToSpeaker = false // set when you begin to record.
// AKSettings.defaultToSpeaker = true // set when you play a recorded file back

do {
     try AVAudioSession.sharedInstance().setCategory(AVAudioSessionCategoryPlayAndRecord,
 with:AKSettings.computedSessionOptions())
} catch {
     print("Filed to change AVAudioSessionCategoryOptions")
}
```
However, with the above trick, it is impossible to record additionally while you are playing a recorded file because calling AVAudioSession.sharedInstance().setCategory() stops a sound that is currently playing.

This noise problem has been issued almost one year ago (https://github.com/audiokit/AudioKit/issues/798) but it doesn't seem to be solved yet. Also, the recent issue (https://github.com/AudioKit/AudioKit/issues/1213) shows the same problem. According to the issue 1213, this mic recording noise problem only appears to happen on devices newer than iPhone 6 (6s and newer).

Is there any solution for it?

**Comment by aure** [comment]
Thanks for submitting this issue in a new form, the old issue was pretty much ignored, hopefully this one will get more attention since its not buried in a sea of other issues.

**Comment by yarivadam** [comment]
Hi!
Any news on this? I'm getting the same clicking noise, and I couldn't make it go away by setting AKSettings.defaultToSpeaker = false before recording (does this AudioKit engine need to be reset for this to take effect?).
Anyway, we'd love to have a solution!
Thanks

**Comment by EssamSoft** [comment]
I’ve avoided this issue by using AVaudiorecorder with AudioKit

**Comment by yarivadam** [comment]
Thanks EssamSoft,
I'm using AudioKit mainly for the FFT function (the AKFFTTap class)... perhaps the solution is getting the FFT data from the mic using AudioKit and recording using AVAudioRecorder. Are there any clashes between using AVFoundation and AudioKit I should be aware of? Should I set the AVAudioSession category in parallel to the AKSettings.setSession(category:) call?
Thanks,
Yariv.

**Comment by jbjbhatcher** [comment]
Hi, 
I was reading the source, specifically: 
https://github.com/AudioKit/AudioKit/blob/master/AudioKit/Common/Internals/AKNotifications.swift
it says that the engine is restarted when the route changes. 
Does anyone think it is possible that the route is continually being switched, resulting in a clicking noise?

**Comment by timrichardsn** [comment]
Hey guys, I'm just wondering if anyone found a workaround for this yet? I've tried using AVAudioRecorder as suggested by @yarivadam, but it hasn't worked with our setup. Anyone else have any ideas, or updates?

**Comment by Dane456** [comment]
+ 1. However I'm not getting clicking I'm getting really bad feedback when I override through `try AVAudioSession.sharedInstance().overrideOutputAudioPort(AVAudioSessionPortOverride.speaker)`. Not to mention the recording audio is getting fed through my earpiece speaker for some reason.

**Comment by eljeff** [comment]
There's a very quick fix for this in the new audiokit. We can make it even better, but I hope it helps for now.

**Comment by aure** [comment]
I'm going to close this for now but if it is not fixed, you can reopen or start a new issue and reference that issue here (preferable).

**Comment by jamfly** [comment]
Hi
I also have this issue for my iPhone 7, however, this issue doesn't happen in my SE.
FYI, My AudioKit version is the latest.
